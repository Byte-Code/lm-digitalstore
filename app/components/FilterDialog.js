import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import glamorous from 'glamorous';
import Toggle from 'material-ui/Toggle';

import FilterButton from './FilterButton';
import Filter from './Filter';

export default class FilterDialog extends Component {
  static propTypes = {
    filterGroups: ImmutablePropTypes.list.isRequired,
    activeFilters: ImmutablePropTypes.list.isRequired,
    initTempFilters: PropTypes.func.isRequired,
    applyTempFilters: PropTypes.func.isRequired,
    toggleTempAvailability: PropTypes.func.isRequired,
    toggleTempFilter: PropTypes.func.isRequired,
    handleClose: PropTypes.func.isRequired,
    availability: PropTypes.bool.isRequired,
    itemCount: PropTypes.number.isRequired
  };

  componentWillMount() {
    this.props.initTempFilters();
  }

  applyAndClose = () => {
    const { applyTempFilters, handleClose } = this.props;
    applyTempFilters();
    handleClose();
  };

  renderFilterGroups = () => {
    const { filterGroups, toggleTempFilter, activeFilters } = this.props;
    return filterGroups.map(g =>
      <GroupWrapper key={g.get('code')}>
        <GroupTitle>{g.get('group')}</GroupTitle>
        <FilterWrapper>
          {g.get('filters').map(f =>
            <Filter
              key={f.get('code')}
              onClick={() => toggleTempFilter(f.get('code'))}
              isActive={activeFilters.contains(f.get('code'))}
            >
              <p>{f.get('name')}</p>
            </Filter>
          )}
        </FilterWrapper>
      </GroupWrapper>
    );
  };

  render() {
    const {
      toggleTempAvailability,
      availability,
      itemCount
    } = this.props;
    const availabilityLabel = availability ? 'Disponibile' : 'Indifferente';
    const labelColor = availability ? '#67cb33' : '#fff';

    return (
      <Wrapper id="filtersDialog">
        <Availability>
          <p>Disponibilit√† in negozio</p>
          <Toggle
            label={availabilityLabel}
            labelPosition="right"
            trackSwitchedStyle={trackSwitchedStyle}
            trackStyle={trackStyle}
            thumbStyle={thumbStyle}
            thumbSwitchedStyle={thumbSwitchedStyle}
            toggled={availability}
            onToggle={toggleTempAvailability}
            labelStyle={labelStyle(labelColor)}
          />
        </Availability>
        <Groups>
          {this.renderFilterGroups()}
        </Groups>
        <FilterButton onApply={this.applyAndClose} itemCount={itemCount} />
      </Wrapper>
    );
  }
}

const Wrapper = glamorous.div({
  display: 'flex',
  flexDirection: 'column',
  color: '#fff',
  backgroundColor: 'rgba(51, 51, 51, 1)'
});

const Groups = glamorous.div({
  margin: '0 40px'
});

const GroupWrapper = glamorous.div({
  width: '100%',
  display: 'flex',
  flexDirection: 'column',
  marginBottom: '10px'
});

const GroupTitle = glamorous.div({
  fontSize: '16px',
  textTransform: 'uppercase',
  marginBottom: '13px',
  fontFamily: 'LeroyMerlinSans Light'
});

const FilterWrapper = glamorous.div({
  display: 'flex',
  flexWrap: 'wrap',
  '&>div': {
    marginRight: '10px',
    marginBottom: 20
  },
  '&>div:last-child': {
    marginRight: 0
  }
});

const Availability = glamorous.div({
  background: 'rgba(103, 203, 51, 0.1)',
  display: 'flex',
  flexDirection: 'column',
  padding: '14px 40px',
  marginBottom: '31px',
  '&>p': {
    fontSize: '16px',
    color: '#fff',
    textTransform: 'uppercase',
    fontFamily: 'LeroyMerlinSans Light',
    marginBottom: '13px'
  }
});

const trackSwitchedStyle = { width: 70, height: 42, backgroundColor: '#fff' };
const trackStyle = { width: 70, height: 42, backgroundColor: '#fff' };
const thumbStyle = { width: 30, height: 30, backgroundColor: '#e4e4e4', top: 10, left: 10 };
const thumbSwitchedStyle = { backgroundColor: '#67cb33', top: 10, left: 65 };
const labelStyle = labelColor => ({
  marginLeft: 50,
  fontSize: 16,
  color: labelColor,
  fontFamily: 'LeroyMerlinSans Italic',
  display: 'flex',
  alignItems: 'center'
});
