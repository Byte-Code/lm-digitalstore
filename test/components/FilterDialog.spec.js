import React from 'react';
import { fromJS } from 'immutable';
import { shallow } from 'enzyme';
import Toggle from 'material-ui/Toggle';

import FilterDialog from '../../app/components/FilterDialog';
import Filter from '../../app/components/Filter';

const applyTempFilters = jest.fn();
const handleClose = jest.fn();
const initTempFilters = jest.fn();
const resetTempFilters = jest.fn();
const toggleTempFilter = jest.fn();
const toggleTempAvailability = jest.fn();
const filterGroups = fromJS([
  {
    code: 'group0',
    filters: [
      { code: 'filter0', products: [0, 1, 2] },
      { code: 'filter1', products: [3, 4, 5] },
      { code: 'filter2', products: [0, 1, 2] }
    ]
  },
  {
    code: 'group1',
    filters: [
      { code: 'filter3', products: [0, 1, 2] },
      { code: 'filter4', products: [3, 4, 9] },
      { code: 'filter5', products: [6, 7, 8] }
    ]
  }
]);
const activeFilters = fromJS(['filter1']);
const result = shallow(
  <FilterDialog
    filterGroups={filterGroups}
    handleClose={handleClose}
    categoryCode="CAT123"
    applyTempFilters={applyTempFilters}
    initTempFilters={initTempFilters}
    toggleTempFilter={toggleTempFilter}
    toggleTempAvailability={toggleTempAvailability}
    resetTempFilters={resetTempFilters}
    activeFilters={activeFilters}
    itemCount={10}
    availability
  />
);

describe('FilterDialog', () => {
  it('should call handleClose when the button is clicked', () => {
    result.instance().applyAndClose();
    expect(handleClose).toHaveBeenCalled();
  });

  it('should call toggleTempFilter when a filter is clicked', () => {
    result.find(Filter).at(0).simulate('click');
    expect(toggleTempFilter).toHaveBeenCalledWith('filter0');
    expect(result).toMatchSnapshot();
  });

  it('should call toggleTempAvailability availability is toggled', () => {
    result.find(Toggle).simulate('toggle');
    expect(toggleTempAvailability).toHaveBeenCalled();
    expect(result).toMatchSnapshot();
  });
});
