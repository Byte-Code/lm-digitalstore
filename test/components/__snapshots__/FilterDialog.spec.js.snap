// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterDialog should call toggleTempAvailability availability is toggled 1`] = `
<glamorous(div)
  id="filtersDialog"
>
  <glamorous(div)>
    <p>
      Disponibilità in negozio
    </p>
    <Toggle
      defaultToggled={false}
      disabled={false}
      label="Disponibile"
      labelPosition="right"
      labelStyle={
        Object {
          "alignItems": "center",
          "color": "#67cb33",
          "display": "flex",
          "fontFamily": "LeroyMerlinSans Italic",
          "fontSize": 16,
          "marginLeft": 50,
        }
      }
      onToggle={[Function]}
      thumbStyle={
        Object {
          "backgroundColor": "#e4e4e4",
          "height": 30,
          "left": 10,
          "top": 10,
          "width": 30,
        }
      }
      thumbSwitchedStyle={
        Object {
          "backgroundColor": "#67cb33",
          "left": 65,
          "top": 10,
        }
      }
      toggled={true}
      trackStyle={
        Object {
          "backgroundColor": "#fff",
          "height": 42,
          "width": 70,
        }
      }
      trackSwitchedStyle={
        Object {
          "backgroundColor": "#fff",
          "height": 42,
          "width": 70,
        }
      }
    />
  </glamorous(div)>
  <glamorous(div)>
    <glamorous(div)>
      <glamorous(div) />
      <glamorous(div)>
        <glamorous(div)
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </glamorous(div)>
        <glamorous(div)
          isActive={true}
          onClick={[Function]}
        >
          <p />
        </glamorous(div)>
        <glamorous(div)
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </glamorous(div)>
      </glamorous(div)>
    </glamorous(div)>
    <glamorous(div)>
      <glamorous(div) />
      <glamorous(div)>
        <glamorous(div)
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </glamorous(div)>
        <glamorous(div)
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </glamorous(div)>
        <glamorous(div)
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </glamorous(div)>
      </glamorous(div)>
    </glamorous(div)>
  </glamorous(div)>
  <FilterButton
    itemCount={10}
    onApply={[Function]}
  />
</glamorous(div)>
`;

exports[`FilterDialog should call toggleTempFilter when a filter is clicked 1`] = `
<glamorous(div)
  id="filtersDialog"
>
  <glamorous(div)>
    <p>
      Disponibilità in negozio
    </p>
    <Toggle
      defaultToggled={false}
      disabled={false}
      label="Disponibile"
      labelPosition="right"
      labelStyle={
        Object {
          "alignItems": "center",
          "color": "#67cb33",
          "display": "flex",
          "fontFamily": "LeroyMerlinSans Italic",
          "fontSize": 16,
          "marginLeft": 50,
        }
      }
      onToggle={[Function]}
      thumbStyle={
        Object {
          "backgroundColor": "#e4e4e4",
          "height": 30,
          "left": 10,
          "top": 10,
          "width": 30,
        }
      }
      thumbSwitchedStyle={
        Object {
          "backgroundColor": "#67cb33",
          "left": 65,
          "top": 10,
        }
      }
      toggled={true}
      trackStyle={
        Object {
          "backgroundColor": "#fff",
          "height": 42,
          "width": 70,
        }
      }
      trackSwitchedStyle={
        Object {
          "backgroundColor": "#fff",
          "height": 42,
          "width": 70,
        }
      }
    />
  </glamorous(div)>
  <glamorous(div)>
    <glamorous(div)>
      <glamorous(div) />
      <glamorous(div)>
        <glamorous(div)
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </glamorous(div)>
        <glamorous(div)
          isActive={true}
          onClick={[Function]}
        >
          <p />
        </glamorous(div)>
        <glamorous(div)
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </glamorous(div)>
      </glamorous(div)>
    </glamorous(div)>
    <glamorous(div)>
      <glamorous(div) />
      <glamorous(div)>
        <glamorous(div)
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </glamorous(div)>
        <glamorous(div)
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </glamorous(div)>
        <glamorous(div)
          isActive={false}
          onClick={[Function]}
        >
          <p />
        </glamorous(div)>
      </glamorous(div)>
    </glamorous(div)>
  </glamorous(div)>
  <FilterButton
    itemCount={10}
    onApply={[Function]}
  />
</glamorous(div)>
`;
